src/coprocessor/dag/rpn_expr/function.rs

```diff
@@ -345,3 +423,171 @@ impl Helper {
+    fn get(&self, row: usize) -> Self::Type;
+}
+
+/// Represents an RPN function argument of a `ScalarValue`.
+#[derive(Clone, Copy, Debug)]
+pub struct ScalarArg<'a, T: Evaluable>(&'a Option<T>);
+
+impl<'a, T: Evaluable> RpnFnArg for ScalarArg<'a, T> {
+    type Type = &'a Option<T>;
+
+    /// Gets the value in the given row. All rows of a `ScalarArg` share the same value.
+    #[inline]
+    fn get(&self, _row: usize) -> &'a Option<T> {
+        self.0
+    }
+}
+
+/// Represents an RPN function argument of a `VectorValue`.
+#[derive(Clone, Copy, Debug)]
+pub struct VectorArg<'a, T: Evaluable>(&'a [Option<T>]);
+
+impl<'a, T: Evaluable> RpnFnArg for VectorArg<'a, T> {
+    type Type = &'a Option<T>;
+
+    #[inline]
+    fn get(&self, row: usize) -> &'a Option<T> {
+        &self.0[row]
+    }
+}
+
+/// Partial or complete argument definition of an RPN function.
+///
+/// `ArgDef` is constructed at the beginning of evaluating an RPN function. The types of
+/// `RpnFnArg`s are determined at this stage. So there won't be dynamic dispatch or enum matches
+/// when the function is applied to each row of the input.
+pub trait ArgDef: std::fmt::Debug {}
+
+/// RPN function argument definitions in the form of a linked list.
+///
+/// For example, if an RPN function foo(Int, Real, Decimal) is applied to input of a scalar of
+/// integer, a vector of reals and a vector of decimals, the constructed `ArgDef` will be
+/// `Arg<ScalarArg<Int>, Arg<VectorValue<Real>, Arg<VectorValue<Decimal>, Null>>>`. `Null`
+/// indicates the end of the argument list.
+#[derive(Debug)]
+pub struct Arg<A: RpnFnArg, Rem: ArgDef> {
+    arg: A,
+    rem: Rem,
+}
+
+impl<A: RpnFnArg, Rem: ArgDef> ArgDef for Arg<A, Rem> {}
+
+impl<A: RpnFnArg, Rem: ArgDef> Arg<A, Rem> {
+    /// Gets the value of the head argument in the given row and returns the remaining argument
+    /// list.
+    #[inline]
+    pub fn extract(&self, row: usize) -> (A::Type, &Rem) {
+        (self.arg.get(row), &self.rem)
+    }
+}
+
+/// Represents the end of the argument list.
+#[derive(Debug)]
+pub struct Null;
+
+impl ArgDef for Null {}
+
+/// A generic evaluator of an RPN function.
+///
+/// For every RPN function, the evaluator should be created first. Then, call its `eval` method
+/// with the input to get the result vector.
+///
+/// There are two kinds of evaluators in general:
+/// - `ArgConstructor`: It's a provided `Evaluator`. It is used in the `rpn_fn` attribute macro
+///   to generate the `ArgDef`. The `def` parameter of its eval method is the already constructed
+///   `ArgDef`. If it is the outmost evaluator, `def` should be `Null`.
+/// - Custom evaluators which do the actual execution of the RPN function. The `def` parameter of
+///   its eval method is the constructed `ArgDef`. Implementors can then extract values from the
+///   arguments, execute the RPN function and fill the result vector.
+pub trait Evaluator {
+    fn eval<D: ArgDef>(
+        self,
+        def: D,
```

Comment:
You could omit all those `D_: ArgDef` in the proc macro if you use `def: impl ArgDef` here 🙃